content/blog/2023-03-17-build-grafana-dashboards-for-apache-skywalking-native-promql-support/index.md

```diff
@@ -0,0 +1,403 @@
+- Service Instance: An individual workload in the Service group.
+- Endpoint: A path in a service for incoming requests.
+- Process: An operating system process. In some scenarios, a Service Instance is not a process, such as a pod Kubernetes could contain multiple processes.
+
+The metric name and properties (labels) are configured by the SkyWalking OAP server based on the data source as well as OAL and MAL.
+SkyWalking provides the ability to down-sampling time series metrics, and generate different time bucket data (minute, hour, day).
+
+The SkyWalking metric stream is as follows:
+
+![SkyWalking metrics stream](img_3.png#pic_left)
+
+Traffic:
+- The metadata of the
+  Service/ServiceRelation/Instance/ServiceInstanceRelation/Endpoint/EndpointRelation/Process/ProcessRelation.
+  Include names, layers, properties, relations between them, etc.
+
+Metric
+- Name: metric name, configuration from OAL and MAL.
+- Entity: represents the metrics' belonging and used for the query.
+  An Entity will contain the following information depending on the `Scope`：
+  Scope represents the metrics level and in query stage represents the Scope catalog,
+  Scope catalog provides high-dimension classifications for all scopes as a hierarchy structure.
+
+| Scope                   | Entity Info                            |
+|-------------------------|----------------------------------------|
+| Service                 | Service(include layer info)            |
+| ServiceInstance         | Service, ServiceInstance               |
+| Endpoint                | Service, Endpoint                      |
+| ServiceRelation         | Service, DestService                   |
+| ServiceInstanceRelation | ServiceInstance, DestServiceInstance   |
+| EndpointRelation        | Endpoint, DestEndpoint                 |
+| Process                 | Service, ServiceInstance, Process      |
+| ProcessRelation         | Process,  ServiceInstance, DestProcess |
+
+- Value:
+1. single value: long.
+2. labeled value: text, `label1,value1|label2,value2|...`, such as `L2 aggregation,5000 | L1 aggregation,8000`.
+
+- TimeBucket: the time is accurate to minute, hour, day.
+
+## How to use PromQL Service
+
+### Setup
+PromQL Service is enabled by default after v9.4.0, so no additional configuration is required.
+The default ports, for example, can be configured by using OAP environment variables:
+```text
+restHost: ${SW_PROMQL_REST_HOST:0.0.0.0}
+restPort: ${SW_PROMQL_REST_PORT:9090}
+restContextPath: ${SW_PROMQL_REST_CONTEXT_PATH:/}
+restMaxThreads: ${SW_PROMQL_REST_MAX_THREADS:200}
+restIdleTimeOut: ${SW_PROMQL_REST_IDLE_TIMEOUT:30000}
+restAcceptQueueSize: ${SW_PROMQL_REST_QUEUE_SIZE:0}
+```
+
+### Use Prometheus expression
+PromQL is matched metric through the Prometheus expression, a typical Prometheus metric is as follows:
+
+![Prometheus metric](img_2.png#pic_left)
+
+
+For the match, the Prometheus expression is as follows:
```

Comment:
These two sentences make me feel a little strange.  We are filtering metrics through this expression. But our metrics should not be 100% same as Prometheus. 

I am thinking how to write this part down.

Reply:
The flight is taking off, I am going to back on this later today.